src/main/java/com/alvarium/hash/exceptions/HashException.java

```diff
@@ -0,0 +1,10 @@
+package com.alvarium.hash.exceptions;
+
+public class HashException {
+  
+  public static class UnsupportedHashType extends Exception {
```

Comment:
I am HUGELY skeptical of this approach. I found the rationale behind using it [here](https://howtodoinjava.com/java/exception-handling/best-practices-for-for-exception-handling/).

1.) I'm personally inclined to avoid `static`types and would probably only use one as a [Singleton](https://en.wikipedia.org/wiki/Singleton_pattern)
2.) The static class shown here doesn't leverage anything at all from its parent class.
3.) You can already organize classes into packages and this technique seems to me just a kind of shorthand for grouping Exception types.

You could just define a `HashTypeException` class and be done with it. I don't expect you'll have a whole range of classes here that would benefit from the additional grouping.

Reply:
fixed in 0d3b63b1f35965acf32b53e9300a6adbc68ea290.

Reply:
@tsconn23 Since your comments here also apply to #18 , I'll discuss things here to keep all relevant concerns in one place.
 
While the only Exception class needed here is only `HashTypeException` as you stated, what would happen if one would need to have more than one Exception class in the same package ? 

My concern is that this approach might clutter the package directory, or is it that if we find ourselves creating a lot of exception classes then we are approaching the issue in an incorrect way and that we should only ever need one Exception class as you said in this case ?  

I suppose we could have only one Exception class and customize the Exception message according to the situation, but I'm looking forward to hearing your take on this.

Reply:
@Ali-Amin There's nothing wrong with having multiple exception types in a given package such as `com.alvarium.hash`. As an example from the JDK, scroll down to the "Exception Summary" on the [java.sql](https://docs.oracle.com/en/java/javase/11/docs/api/java.sql/java/sql/package-summary.html) documentation. There are a whole bunch of exception types in that package. There's even inheritance among those types as shown [here](https://docs.oracle.com/en/java/javase/11/docs/api/java.sql/java/sql/SQLSyntaxErrorException.html).

From what I've seen it's idiomatic in Java to have the Exception classes reside in the package of the functionality they apply to rather than grouped into a separate `exceptions` package. So yes, this could mean you'd have a lot of classes -- both functional and exception -- in the same package.

Reply:
@tsconn23 This approach sounds solid, thanks for clearing this up for me.